src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -323,3 +498,236 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
+        /* when predefined pairtype have larger extent than size, we'll end up
+         * missaligned access. Memcpy the source to workaround the alignment issue.
+         */
+        if (src_kind == MPIDIG_ACC_SRCBUF_PACKED) {
+            src_type_stride = source_dtp_size;
+            if (source_dtp_size < source_dtp_extent) {
+                src_ptr = MPL_malloc(source_dtp_extent, MPL_MEM_OTHER);
+            }
+        } else {
+            src_type_stride = source_dtp_extent;
+        }
+    }
+    if (is_supported) {
+        if (is_empty_source == TRUE || HANDLE_IS_BUILTIN(target_dtp)) {
+            /* directly apply op if target dtp is predefined dtp OR source buffer is empty */
+            max_pack_bytes = source_dtp_size * source_count;
+            mpi_errno =
+                op_pack(source_buf, source_count, source_dtp, 0, target_buf,
+                        max_pack_bytes, &actual_pack_bytes, op);
+
+        } else {
+            /* derived datatype */
+            i = 0;
+            curr_loc = typerep_vec[0].iov_base;
+            curr_len = typerep_vec[0].iov_len;
+            accumulated_count = 0;
+            while (i != vec_len) {
+                if (curr_len < type_size) {
+                    MPIR_Assert(i != vec_len);
+                    i++;
+                    curr_len += typerep_vec[i].iov_len;
+                    continue;
+                }
+
+                MPI_Aint count;
+                MPIR_Assign_trunc(count, curr_len / type_size, MPI_Aint);
+
+                if (src_ptr) {
+                    MPI_Aint unpacked_size;
+                    op_unpack((char *) source_buf +
+                              src_type_stride * accumulated_count, source_dtp_size,
+                              src_ptr, 1, source_dtp, 0, &unpacked_size, op);
+                    MPIR_Typerep_copy((char *) target_buf + MPIR_Ptr_to_aint(curr_loc), src_ptr,
+                                      type_size * count);
+                } else {
+                    max_pack_bytes = type_size * count;
+                    if (source_dtp_size == source_dtp_extent) {
+                        mpi_errno =
+                            op_pack((char *) source_buf +
+                                    src_type_stride * accumulated_count, count,
+                                    type, 0,
+                                    (char *) target_buf +
+                                    MPIR_Ptr_to_aint(curr_loc), max_pack_bytes,
+                                    &actual_pack_bytes, op);
+                    } else {
+                        char *ptr = MPL_malloc(source_dtp_extent, MPL_MEM_OTHER);
+                        MPI_Aint unpacked_size;
+                        mpi_errno =
+                            op_pack((char *) source_buf +
+                                    src_type_stride * accumulated_count, count,
+                                    type, 0, ptr, max_pack_bytes, &actual_pack_bytes, op);
```

Comment:
You are `op_pack` onto a random buffer, that seems wrong.

Reply:
Not sure what you mean by a random buffer here?

Reply:
`ptr` is a new buffer from `MPL_malloc`, which contains random contents, right?

Anyway, I am doing a refactor, will ask you to review shortly.